### PR TITLE
Update code for newer NumPy

### PR DESCRIPTION
**Pull request by oleksandr-pavlyk**
1. Removed use of vendored numpy.pxd, replaced it with cimport numpy to resolve runtime warning during import of change in size of broadcast struct
2. Use explicit namespace for NumPy C-API symbols imported by Cython. This caught several issues where C-API ndarray object was used where Python object of type np.ndarray was intended.
3. Resolve compilation warnings reported by Clang about ignored type qualifier during casting.

Local build now passes, import does not report the runtime warning, and test suite passes.

### CODE REVIEW

File: mkl_random/src/numpy_multiiter_workaround.h [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+#endif
+}
+
+npy_intp* workaround_PyArray_MultiIter_DIMS(PyArrayMultiIterObject *multi) {
+#if defined(NPY_2_0_API_VERSION) && (NPY_API_VERSION >= NPY_2_0_API_VERSION)
+    return PyArray_MultiIter_DIMS(multi);
+#else
+    return (((multi_iter_proxy_st *)(multi))->dimensions);
+#endif
+}
```
Comment (antonwolfy) [inline mkl_random/src/numpy_multiiter_workaround.h:78]: missing empty line at the end of file

File: setup.py [hunk 2]
```diff
@@ -96,7 +96,8 @@ def extensions():
     defs = [('_FILE_OFFSET_BITS', '64'),
             ('_LARGEFILE_SOURCE', '1'),
             ('_LARGEFILE64_SOURCE', '1'),
-            ("PY_ARRAY_UNIQUE_SYMBOL", "mkl_random_ext")]
+            ("PY_ARRAY_UNIQUE_SYMBOL", "mkl_random_ext"),
+            ("NPY_NO_DEPRICATED_API", None)]
```
Comment (antonwolfy) [inline setup.py:100]: Do you mean `NPY_NO_DEPRECATED_API`?
```suggestion
            ("NPY_NO_DEPRECATED_API", None)]
```

File: mkl_random/src/numpy_multiiter_workaround.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+typedef struct {
+    PyObject_HEAD
+    int numiter;
+    npy_intp size;
+    npy_intp index;
+    int nd;
+    npy_intp dimensions[32];
+    void **iters;
+} multi_iter_proxy_st;
```
Comment (antonwolfy) [inline mkl_random/src/numpy_multiiter_workaround.h:54]: Should it be wrapped with if/endif?
```suggestion
#if not defined(NPY_2_0_API_VERSION) || (NPY_API_VERSION < NPY_2_0_API_VERSION)
typedef struct {
    PyObject_HEAD
    int numiter;
    npy_intp size;
    npy_intp index;
    int nd;
    npy_intp dimensions[32];
    void **iters;
} multi_iter_proxy_st;
#endif
```

**Review by antonwolfy — COMMENTED** [review]
Only few minor comments from me.

File: setup.py (hunk 2 above)
Comment (oleksandr-pavlyk) [reply setup.py]: Removed setting the variable altogether. Thank you for spotting the misspelling.

File: mkl_random/src/numpy_multiiter_workaround.h (hunk 1 above)
Comment (oleksandr-pavlyk) [reply mkl_random/src/numpy_multiiter_workaround.h]: Added the end of line to the last line of the file.

File: mkl_random/src/numpy_multiiter_workaround.h (hunk 3 above)
Comment (oleksandr-pavlyk) [reply mkl_random/src/numpy_multiiter_workaround.h:57]: define a preprocessor variable to save repeated evaluations of the same preprocessor expression. Used it to only define this type when necessary.

**Review by xaleryb — APPROVED** [review]

**Pull request merged**